semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Main.scala

```diff
@@ -24,6 +24,36 @@ class Main(settings: Settings, reporter: Reporter) {
 
     val buffer = new ConcurrentLinkedQueue[AbsolutePath]()
 
+    if (!Files.exists(Files.createDirectories(settings.cacheDir.toNIO))) {
+      Files.createDirectories(settings.cacheDir.toNIO)
+    }
+
+    // Writes to a temporary jar file and atomically moves the resulting jar file to cacheEntry once processing
+    // completes. This guarantees the cached jar is never available in half-processed form.
+    def createCachedJar(cacheEntry: AbsolutePath)(f: AbsolutePath => Boolean): Unit = {
+      val tmp = Files.createTempDirectory("metacp").resolve(cacheEntry.toNIO.getFileName)
+      PlatformFileIO.withJarFileSystem(AbsolutePath(tmp), create = true) { out =>
+        val res = f(out)
+        success.compareAndSet(true, res)
+        buffer.add(cacheEntry)
+      }
+      try {
+        Files.move(
+          tmp,
+          cacheEntry.toNIO,
+          StandardCopyOption.ATOMIC_MOVE,
+          StandardCopyOption.REPLACE_EXISTING
+        )
+      } catch {
+        case _: AtomicMoveNotSupportedException =>
```

Comment:
Shall we crash instead?

Reply:
We have no guarantee on which systems ATOMIC_MOVE is not supported, but I would prefer to run successfully on those systems instead of making it impossible to use metacp. Errors from reading half-complete symbols will still be reported.

Reply:
btw, using `move` even without atomic significantly reduces the chances or reproducing the error compared to the previous behavior where we write directly to the target jar while processing.